crates/sui-data-ingestion-core/src/worker_pool.rs

```diff
@@ -118,10 +132,25 @@ impl<W: Worker + 'static> WorkerPool<W> {
                         checkpoints.push_back(checkpoint);
                     } else {
                         let worker_id = idle.pop_first().unwrap();
-                        workers[worker_id].0.send(checkpoint).await.expect("failed to dispatch a task");
+                        if workers[worker_id].0.send(checkpoint).await.is_err() {
+                            // The worker channel closing is a sign we need to exit this loop.
+                            break;
+                        };
                     }
                 }
             }
         }
+
+        // Clean up code for graceful termination
+
+        // Call the termination signal for all workers
+        for (_, term_sender) in workers {
```

Comment:
nit: on drop event, the receiver will be notified that the sender has been dropped. Or if it's solely for awaiting join handles, we can do `drop(workers)` here

Reply:
Nice one, I will use the `drop(workers)`